### ISSUE TITLE
Performance improvements

### ISSUE THREAD

**Issue by mondaychen** [elementalui/elemental#173]
Hi Elemental-UI members,

Do you have a plan to improve the performance of the components?

I'm trying to rebuild my company's UI system based on elemental-ui. However, to be used in a big product, the basic components must have good performance. As far as I could tell, currently elemental-ui components are not optimized by any means.

A simple way to do it is to add a `shouldComponentUpdate` method by default in all components.

Possibly affected components:

```
~/dev/elemental $  ag PropTypes.array
src/components/Dropdown.js
20:		items: React.PropTypes.array.isRequired,

src/components/FormSelect.js
17:		options: React.PropTypes.arrayOf(

src/components/RadioGroup.js
13:		options: React.PropTypes.array.isRequired,

src/components/SegmentedControl.js
11:		options: React.PropTypes.array.isRequired,

~/dev/elemental $  ag PropTypes.object
site/src/components/DemoBox.js
17:		style: React.PropTypes.object,

src/components/Col.js
15:		style: React.PropTypes.object,

src/components/Checkbox.js
13:		style: React.PropTypes.object,

src/components/Container.js
12:		style: React.PropTypes.object,

src/components/FileUpload.js
13:		file: React.PropTypes.object, // https://developer.mozilla.org/en/docs/Using_files_from_web_applications

src/components/FormLabel.js
11:		style: React.PropTypes.object,

src/components/Pagination.js
39:		style: React.PropTypes.object,

src/components/Row.js
12:		style: React.PropTypes.object,
```

**Comment by mondaychen** [comment]
A related issue: https://github.com/elementalui/elemental/issues/46

The validation process is actually not slow. It's slow because the whole page is re-rendered when the state (value of the input fields) changes.

**Comment by jossmac** [comment]
@mondaychen thanks for your contribution.

Might it be easier to implement [`PureComponent`](https://facebook.github.io/react/docs/react-api.html#react.purecomponent) rather than `shouldComponentUpdate`?

Any chance we could get a PR for this?

**Comment by mondaychen** [comment]
Of course, `PureComponent` will be easier. But it was not there yet when I created this issue :)

According to [this tweet](https://twitter.com/reactjs/status/759099486513704961?lang=en), it was shipped with v15.3.0. Do you plan to drop support for older version React, or maybe we can use a polyfill for it?

**Comment by jossmac** [comment]
Haha fair enough. Yeah, I think it's time this project was updated -- it'd be a bump to major, we wouldn't force `>= 15.x` on consumers.
